15th January 2019 

Machine Learning Lab

1. Extracting all rows and columns from the given dataset
2. Extracting from row number 0 to row number 10, column number 0 to column number 2
3. Extracting from row number 12 to last row, column number 1 to last column
4. Dividing your dataset into two parts: independent variables and dependent variable
5. Apply data pre-processing steps:
    Missing Values (Mean, Median & Most Frequent) [Imputer]
    Labelling Categorical Data through One-Hot-Encoding [LabelEncoder, OneHotEncoder]
    Scaling Data [StandardScalar]

In [1]:
# Checking Current Working Directory
import os
print(os.getcwd())

D:\Amity University\6th Semester\Machine Learning Lab\15th January


In [2]:
# Importing Libraries
import pandas as pd
import numpy as np

In [3]:
dataset = pd.read_csv('Data1.csv')
print(dataset.head())

   R&D Spend  Administration  Marketing Spend       State     Profit
0  165349.20       136897.80              NaN    New York  192261.83
1  162597.70       151377.59        443898.53  California  191792.06
2  153441.51       101145.55        407934.54     Florida  191050.39
3  144372.41       118671.85        383199.62    New York  182901.99
4  142107.34        91391.77        366168.42     Florida  166187.94


Independent Variables: R&D Spend, Administration, Marketing Spend, State

Dependent Variable: Profit

In [4]:
# Extracting all rows and columns from the given dataset
dataset_1 = dataset.iloc[:,:]
print(dataset_1)

    R&D Spend  Administration  Marketing Spend       State     Profit
0   165349.20       136897.80              NaN    New York  192261.83
1   162597.70       151377.59        443898.53  California  191792.06
2   153441.51       101145.55        407934.54     Florida  191050.39
3   144372.41       118671.85        383199.62    New York  182901.99
4   142107.34        91391.77        366168.42     Florida  166187.94
5   131876.90        99814.71        362861.36    New York  156991.12
6         NaN       147198.87        127716.82  California  156122.51
7   130298.13       145530.06        323876.68     Florida  155752.60
8   120542.52       148718.95        311613.29    New York  152211.77
9   123334.88       108679.17        304981.62  California  149759.96
10  101913.08       110594.11        229160.95     Florida  146121.95
11  100671.96        91790.61        249744.55  California  144259.40
12   93863.75       127320.38        249839.44     Florida  141585.52
13   91992.39       

In [5]:
# Extracting from Row Number 0 to Row Number 10
# Extracting From Column Number 0 to Column Number 2
dataset_2 = dataset.iloc[:11,:3]
print(dataset_2)

    R&D Spend  Administration  Marketing Spend
0   165349.20       136897.80              NaN
1   162597.70       151377.59        443898.53
2   153441.51       101145.55        407934.54
3   144372.41       118671.85        383199.62
4   142107.34        91391.77        366168.42
5   131876.90        99814.71        362861.36
6         NaN       147198.87        127716.82
7   130298.13       145530.06        323876.68
8   120542.52       148718.95        311613.29
9   123334.88       108679.17        304981.62
10  101913.08       110594.11        229160.95


In [6]:
# Extracting from Row Number 12 to Last Row
# Extracting from Column Number 1 to Last Column
dataset_3 = dataset.iloc[12:,1:]
print(dataset_3)

    Administration  Marketing Spend       State     Profit
12       127320.38        249839.44     Florida  141585.52
13       135495.07        252664.93  California  134307.35
14       156547.42        256512.92     Florida  132602.65
15       122616.84        261776.23    New York  129917.04
16       121597.55        264346.06  California  126992.93
17       145077.58        282574.31    New York  125370.37
18       114175.79        294919.57     Florida  124266.90
19       153514.11             0.00    New York  122776.86
20       113867.30              NaN  California  118474.03
21       153773.43        299737.29    New York  111313.02
22       122782.75        303319.26     Florida  110352.25
23       105751.03        304768.73     Florida  108733.99
24        99281.34        140574.81    New York  108552.04
25       139553.16        137962.62  California  107404.34
26       144135.98        134050.07     Florida  105733.54
27       127864.55        353183.81    New York  105008.

In [7]:
# Dividing dataset into independent variables and dependent variables
x = dataset.iloc[:,:-1].values
y = dataset.iloc[:,-1].values
print(x)
print(y)

[[165349.2 136897.8 nan 'New York']
 [162597.7 151377.59 443898.53 'California']
 [153441.51 101145.55 407934.54 'Florida']
 [144372.41 118671.85 383199.62 'New York']
 [142107.34 91391.77 366168.42 'Florida']
 [131876.9 99814.71 362861.36 'New York']
 [nan 147198.87 127716.82 'California']
 [130298.13 145530.06 323876.68 'Florida']
 [120542.52 148718.95 311613.29 'New York']
 [123334.88 108679.17 304981.62 'California']
 [101913.08 110594.11 229160.95 'Florida']
 [100671.96 91790.61 249744.55 'California']
 [93863.75 127320.38 249839.44 'Florida']
 [91992.39 135495.07 252664.93 'California']
 [119943.24 156547.42 256512.92 'Florida']
 [114523.61 122616.84 261776.23 'New York']
 [78013.11 121597.55 264346.06 'California']
 [94657.16 145077.58 282574.31 'New York']
 [91749.16 114175.79 294919.57 'Florida']
 [86419.7 153514.11 0.0 'New York']
 [76253.86 113867.3 nan 'California']
 [78389.47 153773.43 299737.29 'New York']
 [73994.56 122782.75 303319.26 'Florida']
 [67532.53 105751.03 304

In [8]:
# Data Pre-Processing: Missing Values
from sklearn.preprocessing import Imputer
impute = Imputer(strategy="mean", axis=0) # Default Strategy: Mean
x[:, 0:3] = impute.fit_transform(x[:,0:3])
print(x)

[[165349.2 136897.8 212626.22434782612 'New York']
 [162597.7 151377.59 443898.53 'California']
 [153441.51 101145.55 407934.54 'Florida']
 [144372.41 118671.85 383199.62 'New York']
 [142107.34 91391.77 366168.42 'Florida']
 [131876.9 99814.71 362861.36 'New York']
 [75769.8652173913 147198.87 127716.82 'California']
 [130298.13 145530.06 323876.68 'Florida']
 [120542.52 148718.95 311613.29 'New York']
 [123334.88 108679.17 304981.62 'California']
 [101913.08 110594.11 229160.95 'Florida']
 [100671.96 91790.61 249744.55 'California']
 [93863.75 127320.38 249839.44 'Florida']
 [91992.39 135495.07 252664.93 'California']
 [119943.24 156547.42 256512.92 'Florida']
 [114523.61 122616.84 261776.23 'New York']
 [78013.11 121597.55 264346.06 'California']
 [94657.16 145077.58 282574.31 'New York']
 [91749.16 114175.79 294919.57 'Florida']
 [86419.7 153514.11 0.0 'New York']
 [76253.86 113867.3 212626.22434782612 'California']
 [78389.47 153773.43 299737.29 'New York']
 [73994.56 122782.75 30

Mean Value for Column 0: 75769.86521739131

Mean Value for Column 1: 121228.07693877546

Mean Value for Column 2: 212626.22434782612

In [9]:
impute = Imputer(strategy="median", axis=0) # Strategy: Median
x[:, 0:3] = impute.fit_transform(x[:,0:3])
print(x)

[[165349.2 136897.8 212626.22434782612 'New York']
 [162597.7 151377.59 443898.53 'California']
 [153441.51 101145.55 407934.54 'Florida']
 [144372.41 118671.85 383199.62 'New York']
 [142107.34 91391.77 366168.42 'Florida']
 [131876.9 99814.71 362861.36 'New York']
 [75769.8652173913 147198.87 127716.82 'California']
 [130298.13 145530.06 323876.68 'Florida']
 [120542.52 148718.95 311613.29 'New York']
 [123334.88 108679.17 304981.62 'California']
 [101913.08 110594.11 229160.95 'Florida']
 [100671.96 91790.61 249744.55 'California']
 [93863.75 127320.38 249839.44 'Florida']
 [91992.39 135495.07 252664.93 'California']
 [119943.24 156547.42 256512.92 'Florida']
 [114523.61 122616.84 261776.23 'New York']
 [78013.11 121597.55 264346.06 'California']
 [94657.16 145077.58 282574.31 'New York']
 [91749.16 114175.79 294919.57 'Florida']
 [86419.7 153514.11 0.0 'New York']
 [76253.86 113867.3 212626.22434782612 'California']
 [78389.47 153773.43 299737.29 'New York']
 [73994.56 122782.75 30

Median Value for Column 0: 74661.715

Median Value for Column 1: 122616.84

Median Value for Column 2: 212716.24


In [10]:
impute = Imputer(strategy="most_frequent", axis=0) # Strategy: Most Frequent
x[:, 0:3] = impute.fit_transform(x[:,0:3])
print(x)

[[165349.2 136897.8 212626.22434782612 'New York']
 [162597.7 151377.59 443898.53 'California']
 [153441.51 101145.55 407934.54 'Florida']
 [144372.41 118671.85 383199.62 'New York']
 [142107.34 91391.77 366168.42 'Florida']
 [131876.9 99814.71 362861.36 'New York']
 [75769.8652173913 147198.87 127716.82 'California']
 [130298.13 145530.06 323876.68 'Florida']
 [120542.52 148718.95 311613.29 'New York']
 [123334.88 108679.17 304981.62 'California']
 [101913.08 110594.11 229160.95 'Florida']
 [100671.96 91790.61 249744.55 'California']
 [93863.75 127320.38 249839.44 'Florida']
 [91992.39 135495.07 252664.93 'California']
 [119943.24 156547.42 256512.92 'Florida']
 [114523.61 122616.84 261776.23 'New York']
 [78013.11 121597.55 264346.06 'California']
 [94657.16 145077.58 282574.31 'New York']
 [91749.16 114175.79 294919.57 'Florida']
 [86419.7 153514.11 0.0 'New York']
 [76253.86 113867.3 212626.22434782612 'California']
 [78389.47 153773.43 299737.29 'New York']
 [73994.56 122782.75 30

Most Frequent Value for Column 0: 542.05

Most Frequent Value for Column 1: 51283.14

Most Frequent Value for Column 2: 0.0

In [11]:
# Labelling Categorical Data
from sklearn.preprocessing import LabelEncoder
le = LabelEncoder()
x[:, 3] = le.fit_transform(x[:, 3])
print(le.classes_)

['California' 'Florida' 'New York']


Classes of Label Encoder: ['California' 'Florida' 'New York']

Mapping: {‘California’: 0, ‘Florida’: 1, ‘New York’:2}

In [12]:
# Performing One Hot Encoding
from sklearn.preprocessing import OneHotEncoder
enc = OneHotEncoder()
countries = enc.fit_transform(x[:, 3].reshape(-1, 1)).toarray()
x = x[:, :-1]
x = np.concatenate((x, countries), axis=1)
print(x)

[[165349.2 136897.8 212626.22434782612 0.0 0.0 1.0]
 [162597.7 151377.59 443898.53 1.0 0.0 0.0]
 [153441.51 101145.55 407934.54 0.0 1.0 0.0]
 [144372.41 118671.85 383199.62 0.0 0.0 1.0]
 [142107.34 91391.77 366168.42 0.0 1.0 0.0]
 [131876.9 99814.71 362861.36 0.0 0.0 1.0]
 [75769.8652173913 147198.87 127716.82 1.0 0.0 0.0]
 [130298.13 145530.06 323876.68 0.0 1.0 0.0]
 [120542.52 148718.95 311613.29 0.0 0.0 1.0]
 [123334.88 108679.17 304981.62 1.0 0.0 0.0]
 [101913.08 110594.11 229160.95 0.0 1.0 0.0]
 [100671.96 91790.61 249744.55 1.0 0.0 0.0]
 [93863.75 127320.38 249839.44 0.0 1.0 0.0]
 [91992.39 135495.07 252664.93 1.0 0.0 0.0]
 [119943.24 156547.42 256512.92 0.0 1.0 0.0]
 [114523.61 122616.84 261776.23 0.0 0.0 1.0]
 [78013.11 121597.55 264346.06 1.0 0.0 0.0]
 [94657.16 145077.58 282574.31 0.0 0.0 1.0]
 [91749.16 114175.79 294919.57 0.0 1.0 0.0]
 [86419.7 153514.11 0.0 0.0 0.0 1.0]
 [76253.86 113867.3 212626.22434782612 1.0 0.0 0.0]
 [78389.47 153773.43 299737.29 0.0 0.0 1.0]
 [73994.


Mapping: {2:[0,0,1], 0:[1,0,0],1:[0,1,0]}


In [13]:
# Reducing Dimensions
# (50, 6) -> (50, 5)
x = x[:,:-1]
print(x)

[[165349.2 136897.8 212626.22434782612 0.0 0.0]
 [162597.7 151377.59 443898.53 1.0 0.0]
 [153441.51 101145.55 407934.54 0.0 1.0]
 [144372.41 118671.85 383199.62 0.0 0.0]
 [142107.34 91391.77 366168.42 0.0 1.0]
 [131876.9 99814.71 362861.36 0.0 0.0]
 [75769.8652173913 147198.87 127716.82 1.0 0.0]
 [130298.13 145530.06 323876.68 0.0 1.0]
 [120542.52 148718.95 311613.29 0.0 0.0]
 [123334.88 108679.17 304981.62 1.0 0.0]
 [101913.08 110594.11 229160.95 0.0 1.0]
 [100671.96 91790.61 249744.55 1.0 0.0]
 [93863.75 127320.38 249839.44 0.0 1.0]
 [91992.39 135495.07 252664.93 1.0 0.0]
 [119943.24 156547.42 256512.92 0.0 1.0]
 [114523.61 122616.84 261776.23 0.0 0.0]
 [78013.11 121597.55 264346.06 1.0 0.0]
 [94657.16 145077.58 282574.31 0.0 0.0]
 [91749.16 114175.79 294919.57 0.0 1.0]
 [86419.7 153514.11 0.0 0.0 0.0]
 [76253.86 113867.3 212626.22434782612 1.0 0.0]
 [78389.47 153773.43 299737.29 0.0 0.0]
 [73994.56 122782.75 303319.26 0.0 1.0]
 [67532.53 105751.03 304768.73 0.0 1.0]
 [77044.01 99281

In [14]:
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
x = scaler.fit_transform(x)
print(x)

[[ 2.13025600e+00  5.65200076e-01  0.00000000e+00 -7.17740563e-01
  -6.85994341e-01]
 [ 2.06482350e+00  1.08747979e+00  2.16999971e+00  1.39326109e+00
  -6.85994341e-01]
 [ 1.84708323e+00 -7.24367988e-01  1.83255400e+00 -7.17740563e-01
   1.45773797e+00]
 [ 1.63141402e+00 -9.22019908e-02  1.60046927e+00 -7.17740563e-01
  -6.85994341e-01]
 [ 1.57754915e+00 -1.07618258e+00  1.44066761e+00 -7.17740563e-01
   1.45773797e+00]
 [ 1.33426250e+00 -7.72370806e-01  1.40963787e+00 -7.17740563e-01
  -6.85994341e-01]
 [-3.46054196e-16  9.36755178e-01 -7.96694538e-01  1.39326109e+00
  -6.85994341e-01]
 [ 1.29671830e+00  8.76561930e-01  1.04384939e+00 -7.17740563e-01
   1.45773797e+00]
 [ 1.06472343e+00  9.91583801e-01  9.28783509e-01 -7.17740563e-01
  -6.85994341e-01]
 [ 1.13112760e+00 -4.52633600e-01  8.66559362e-01  1.39326109e+00
  -6.85994341e-01]
 [ 6.21702987e-01 -3.83562549e-01  1.55143305e-01 -7.17740563e-01
   1.45773797e+00]
 [ 5.92188330e-01 -1.06179659e+00  3.48276702e-01  1.39326109e+00

D:\Anaconda3\lib\site-packages\sklearn\utils\validation.py:475: DataConversionWarning: Data with input dtype object was converted to float64 by StandardScaler.
  warnings.warn(msg, DataConversionWarning)



Mean of Every Column = array([-3.88837804e-16,  2.03670414e-15, -3.37507799e-16, -1.77635684e-17, -1.11022302e-16])

Standard Deviation of Every Column = array([1., 1., 1., 1., 1.])
